File: App/Sources/Design/ReferenceSheet.swift
import SwiftUI

/// A page of reading matter: rows a learner runs their eye down and taps to hear,
/// gathered into titled panels. The numbers table, the atlas and the calendar are
/// this one sheet with three sets of rows.
///
/// The sheet owns the ASKING side of the page — the heading, the tap hint and
/// whether it is offered at all, the panels and their names, and the fact that one
/// row is one stop that says its own learned side. What stands INSIDE a row is the
/// caller's, because that is the only place the three differ: a flag and two
/// columns, a name and the forms under it, a numeral and its reading.
///
/// [speak] decides the hint as well as the rows: where the device can say nothing
/// on the page, the hint would promise a gesture that does nothing.
struct ReferenceSheet<Row, Content: View>: View {
    /// The page's own heading, where the sheet carries one. The numbers table is
    /// shown under the overview's heading and again in a sheet of its own, so it
    /// supplies none and lets each door title it.
    var heading: LocalizedStringKey?
    let groups: [ReferenceGroup<Row>]
    /// Saying one row in the language being learned — nil where the device can
    /// neither play nor speak it.
    let speak: (Row) -> (() -> Void)?
    /// A group's rows as they stand inside the panel. `ReferenceRows` is what a
    /// plain column of them looks like; the numbers table pairs its columns.
    @ViewBuilder let panel: ([Row]) -> Content

    var body: some View {
        VStack(alignment: .leading, spacing: Theme.spacing.lg) {
            if let heading {
                DrillHeading(heading)
            }
            if groups.contains(where: audible) {
                ReferenceTapHint()
            }
            ForEach(Array(groups.enumerated()), id: \.offset) { _, group in
                VStack(alignment: .leading, spacing: Theme.spacing.sm) {
                    if let title = group.title {
                        Text(title)
                            .font(Theme.typography.caption)
                            .foregroundStyle(Theme.colors.textSecondary)
                            .textCase(.uppercase)
                            .accessibilityAddTraits(.isHeader)
                    }
                    panel(group.rows)
                        .panelSurface()
                }
            }
        }
    }

    private func audible(_ group: ReferenceGroup<Row>) -> Bool {
        group.rows.contains { speak($0) != nil }
    }
}

extension ReferenceSheet {
    /// The plain sheet: one tappable row per entry, read straight down its panel.
    init<RowView: View>(heading: LocalizedStringKey? = nil,
                        groups: [ReferenceGroup<Row>],
                        speak: @escaping (Row) -> (() -> Void)?,
                        @ViewBuilder row: @escaping (Row) -> RowView)
    where Content == ReferenceRows<Row, RowView> {
        self.init(heading: heading, groups: groups, speak: speak,
                  panel: { ReferenceRows(rows: $0, speak: speak, row: row) })
    }
}

/// One panel of a reference sheet: the rows standing in it, and what the block is
/// called. A group goes unnamed where the sheet has only one kind of row, or where
/// this build has no wording for a band kern has just grown.
struct ReferenceGroup<Row> {
    var title: LocalizedStringKey?
    let rows: [Row]
}

/// A panel's rows, one under the other, each saying its own learned side.
///
/// The WHOLE row is the target, and nothing on the row says so: reading matter is
/// read by running down it, and a glyph to aim at is a detour per line. The sheet's
/// hint discloses the gesture once, for the page.
struct ReferenceRows<Row, RowView: View>: View {
    let rows: [Row]
    /// How far apart the lines stand. A band of counting words is read at a glance
    /// and sits tighter than a table of two-column entries.
    var spacing: CGFloat = Theme.spacing.lg
    let speak: (Row) -> (() -> Void)?
    @ViewBuilder let row: (Row) -> RowView

    var body: some View {
        VStack(alignment: .leading, spacing: spacing) {
            ForEach(Array(rows.enumerated()), id: \.offset) { _, entry in
                row(entry)
                    // why: one row is one VoiceOver stop — both sides of the table
                    // are the same entry, not two the reader has to pair up by ear.
                    .accessibilityElement(children: .combine)
                    .pronounceOnTap(speak(entry))
            }
        }
    }
}

/// What tells a reference page's reader that the rows sound: on a page of reading
/// matter the content is the control, so the gesture is disclosed once for the
/// whole page instead of by a glyph on every row. Drawn by the sheets and the box,
/// and only where the device can say the language.
///
/// Silent to VoiceOver: every row already offers hearing it as an action, and a
/// line that exists to be seen is noise when it is read out.
struct ReferenceTapHint: View {
    /// Defaults to the reference tables' own wording; the box names its rows
    /// "words" rather than a table's, so it passes its own key.
    var textKey: LocalizedStringKey = "trainer.reference.tapToHear"

    var body: some View {
        HStack(spacing: Theme.spacing.sm) {
            Image(systemName: "speaker.wave.2.fill")
            Text(textKey)
        }
        .font(Theme.typography.caption)
        .foregroundStyle(Theme.colors.textSecondary)
        .accessibilityHidden(true)
    }
}

/// The authored prose beside a generated table: two to four lines the catalog
/// carries per language, picked for the reader — kern falls back to English where
/// their own language carries no wording.
///
/// A generated table can say nothing its rows do not carry, so how the language
/// ASSEMBLES what it counts, and what trips a learner up doing it, is written
/// rather than derived.
struct ReferenceNotes: View {
    let lines: [String]

    @ViewBuilder
    var body: some View {
        if !lines.isEmpty {
            VStack(alignment: .leading, spacing: Theme.spacing.lg) {
                DrillHeading("common.notes")
                VStack(alignment: .leading, spacing: Theme.spacing.md) {
                    ForEach(Array(lines.enumerated()), id: \.offset) { _, line in
                        HStack(alignment: .firstTextBaseline, spacing: Theme.spacing.sm) {
                            Text(verbatim: "·")
                                .foregroundStyle(Theme.colors.textSecondary)
                                .accessibilityHidden(true)
                            Text(verbatim: line)
                                .font(Theme.typography.subheadline)
                                .foregroundStyle(Theme.colors.textPrimary)
                                .fixedSize(horizontal: false, vertical: true)
                        }
                    }
                }
                .panelSurface()
            }
        }
    }
}

File: android/src/main/kotlin/net/spross/app/ui/ReferenceSheet.kt
package net.spross.app.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import net.spross.app.AppModel
import net.spross.app.Chrome
import net.spross.app.speakFormOnTap
import net.spross.kern.model.Language

/**
 * The reading half of a drill page: the table the run grades against, written down — every
 * entry the pair joins, both sides beside each other, in captioned blocks.
 *
 * Reading matter, so the CONTENT is the control: a whole row says its learned-side name, and
 * the page discloses that gesture once under its heading instead of growing a speaker on
 * every line.
 *
 * Which table, how its blocks are captioned and what each row carries under a name is the
 * page's — the atlas, the calendar — and the shape they are read in is this.
 */

/** One captioned block of a sheet: the rows standing at one tier, or one pool of names. */
class ReferenceGroup(val caption: String, val rows: List<ReferenceRow>)

/** One entry of a sheet, both sides beside each other. */
class ReferenceRow(
    val source: String,
    val target: String,
    /** The picture leading the row; null where the table has none to show. */
    val emoji: String? = null,
    /** What stands under the known name — null where the row has nothing to add there. */
    val sourceUnder: String? = null,
    /** What stands under the learned name — the forms it also answers to. */
    val targetUnder: String? = null,
)

@Composable
fun ReferenceSection(
    model: AppModel,
    heading: String,
    groups: List<ReferenceGroup>,
    source: Language,
    target: Language,
    chrome: Chrome,
) {
    // Heading and hint stand in the page's own rhythm rather than a Column of their own,
    // so the line sits the same distance under its heading as it does on the numbers page.
    OverviewHeading(heading)
    // The hint is the affordance: where nothing on the page can be heard it would
    // promise a gesture that does nothing.
    val audible = groups.any { group ->
        group.rows.any { model.speakFormOnTap(it.target, target) != null }
    }
    if (audible) TapToHearHint(chrome)
    for (group in groups) {
        Column(verticalArrangement = Arrangement.spacedBy(Theme.spacing.sm)) {
            Text(
                group.caption.uppercase(),
                style = MaterialTheme.typography.bodySmall,
                color = Theme.colors.textSecondary,
                modifier = Modifier.semantics { heading() },
            )
            OverviewPanel {
                for (row in group.rows) SheetRow(row, model, source, target, chrome)
            }
        }
    }
}

/**
 * One entry, twice: the known language on the left, the learned one on the right, each with
 * whatever the table adds under the name.
 *
 * The whole row says the LEARNED side. The other column is the reader's own language, and a
 * reference sheet is read to hear what one cannot yet say.
 */
@Composable
private fun SheetRow(
    row: ReferenceRow,
    model: AppModel,
    source: Language,
    target: Language,
    chrome: Chrome,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            // why: one entry is one TalkBack stop — both names and whatever stands under
            // them are the same row of the table.
            .semantics(mergeDescendants = true) { }
            .pronounceOnTap(
                model.speakFormOnTap(row.target, target),
                chrome,
                minHeight = 0.dp,
            ),
        verticalAlignment = Alignment.Top,
        horizontalArrangement = Arrangement.spacedBy(Theme.spacing.md),
    ) {
        row.emoji?.let { Text(it, fontSize = 28.sp, modifier = Modifier.clearAndSetSemantics { }) } // card-parity: a picture, not a type role
        SheetSide(
            name = row.source,
            under = row.sourceUnder,
            language = source,
            tint = Theme.colors.textPrimary,
            align = TextAlign.Start,
            alignment = Alignment.Start,
            modifier = Modifier.weight(1f),
        )
        SheetSide(
            name = row.target,
            under = row.targetUnder,
            language = target,
            tint = Theme.colors.accent,
            align = TextAlign.End,
            alignment = Alignment.End,
            modifier = Modifier.weight(1f),
        )
    }
}

/** [language] tags the column, so TalkBack reads each side in its own voice. */
@Composable
private fun SheetSide(
    name: String,
    under: String?,
    language: Language,
    tint: Color,
    align: TextAlign,
    alignment: Alignment.Horizontal,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(2.dp), // card-parity: the line under a name sits tighter than xs
        horizontalAlignment = alignment,
    ) {
        Text(
            localizedTarget(name, language),
            style = MaterialTheme.typography.titleMedium,
            color = tint,
            textAlign = align,
        )
        under?.let {
            Text(
                localizedTarget(it, language),
                style = MaterialTheme.typography.bodySmall,
                color = Theme.colors.textSecondary,
                textAlign = align,
            )
        }
    }
}

/**
 * What a sheet's rows cannot say: how the language assembles what it lists and what trips a
 * learner up doing it, written in the known language. Nothing at all where the pair carries none.
 */
@Composable
fun ReferenceNotes(lines: List<String>, chrome: Chrome) {
    if (lines.isEmpty()) return
    OverviewHeading(chrome.commonNotes)
    OverviewPanel {
        Column(verticalArrangement = Arrangement.spacedBy(Theme.spacing.md)) {
            for (line in lines) {
                Row(horizontalArrangement = Arrangement.spacedBy(Theme.spacing.sm)) {
                    Text(
                        "·",
                        style = MaterialTheme.typography.bodyMedium,
                        color = Theme.colors.textSecondary,
                        modifier = Modifier.clearAndSetSemantics { },
                    )
                    Text(line, style = MaterialTheme.typography.bodyMedium)
                }
            }
        }
    }
}
